conn_test.go

```diff
@@ -256,6 +257,163 @@ func waitSignal(sigc <-chan *Signal, name string, timeout time.Duration) *Signal
 	}
 }
 
+const (
+	SCPPInterface         = "org.godbus.DBus.StatefulTest"
+	SCPPPath              = "/org/godbus/DBus/StatefulTest"
+	SCPPChangedSignalName = "Changed"
+	SCPPStateMethodName   = "State"
+)
+
+func TestStateCachingProxyPattern(t *testing.T) {
+	srv, err := ConnectSessionBus()
+	defer srv.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	conn, err := ConnectSessionBus(WithSignalHandler(NewSequentialSignalHandler()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer conn.Close()
+
+	serviceName := srv.Names()[0]
+	messages := make(chan *Message)
+	srv.Eavesdrop(messages)
```

Comment:
Why Eavesdrop? This isn't the normal way to process messages for servers.

Reply:
In this test, we have a server which exports an object with a state that clients can query over DBus and subscribe to state change notifications for. 

For this test, it is important that the server's outputs as sent over DBus (its emitted signals and method replies) provide a consistent view of the object's state. That is, each DBus message sent should reflect the state at some point in time, and that time should be monotonically increasing with each DBus message. I think this is basically Tanenbaum's "Monotonic read consistency".

If we are handling a request for the current state around the same time a few state change events are occuring, the following outputs would all be valid (S represents a state change signal, R a response to a method call querying the current state):
- S(38->39), R(39), S(39->40) 
- R(38), S(38->39), S(39->40) 
- S(38->39), S(39->40), R(40)

These, on the other hand, would not be valid:
S(38->39), S(39->40), R(39)
S(38->39), S(39->40), R(38)
S(38->39), R(40), S(39->40)
etc.

This appears to difficult to achieve if I simply export() an object to get the library handle method replies for me and use a second goroutine to emit events on a timer. 
If I use export(), each method request will end up calling a function I designate, but there is no way I can prevent the second goroutine from advancing the state and emitting another event _before_ my method response hits the dbus. Currently, the first goroutine to get to .sendMessageAndIfClosed() gets to send its message first, and this has nothing to do with which goroutine read the application-level state first.

Reply:
In summary, the issue I'm working around here with evesdrop is basically the server-side counterpart to the client-side issue this patch is fixing. 

For our use case, this is not really a concern as the server is implemented in a different language and (consequently) does not use this library.